apollo-router/src/plugins/cors.rs

```diff
@@ -0,0 +1,572 @@
+                &config,
+                &policy,
+                &request_origin,
+                true,
+                request_headers,
+            );
+            return Box::pin(async move { Ok(response) });
+        }
+
+        let fut = self.inner.call(req);
+        Box::pin(async move {
+            let mut response = fut.await?;
+            // Find matching origin configuration
+            let policy = Self::find_matching_policy(&config, &request_origin);
+            // Add CORS headers for non-preflight
+            Self::add_cors_headers(
+                &mut response,
+                &config,
+                &policy,
+                &request_origin,
+                false,
+                request_headers,
+            );
+            Ok(response)
+        })
+    }
+}
+
+impl<S> CorsService<S> {
+    /// Find the matching policy for a given origin
+    fn find_matching_policy<'a>(
+        config: &'a Cors,
+        origin: &'a Option<http::HeaderValue>,
+    ) -> Option<&'a Policy> {
+        let origin_str = origin.as_ref()?.to_str().ok()?;
+
+        if let Some(policies) = &config.policies {
+            for policy in policies.iter() {
+                for url in &policy.origins {
+                    if url == origin_str {
+                        return Some(policy);
+                    }
+                }
+
+                if !policy.match_origins.is_empty() {
+                    for regex in &policy.match_origins {
+                        if regex.is_match(origin_str) {
+                            return Some(policy);
+                        }
+                    }
+                }
+            }
+        }
+
+        None
+    }
+
+    /// Add CORS headers to the response
+    fn add_cors_headers<ResBody>(
+        response: &mut Response<ResBody>,
+        config: &Cors,
+        policy: &Option<&Policy>,
+        request_origin: &Option<http::HeaderValue>,
+        is_preflight: bool,
+        request_headers: Option<http::HeaderValue>,
+    ) {
+        let allow_credentials = policy
+            .and_then(|p| p.allow_credentials)
+            .unwrap_or(config.allow_credentials);
+
+        let allow_headers = policy
+            .and_then(|p| {
+                if p.allow_headers.is_empty() {
+                    None
+                } else {
+                    Some(&p.allow_headers)
+                }
+            })
+            .unwrap_or(&config.allow_headers);
+
+        // Distinguish between None, Some([]), and Some([item, ...]) for expose_headers
+        let expose_headers = if let Some(policy) = policy {
+            if policy.expose_headers.is_empty() {
+                config.expose_headers.as_ref()
+            } else {
+                Some(&policy.expose_headers)
+            }
+        } else {
+            config.expose_headers.as_ref()
+        };
+
+        let methods = policy
+            .and_then(|p| {
+                if p.methods.is_empty() {
```

Comment:
Per my comment in config... the default for this field is currently not empty or None, but  GET, POST,  OPTIONS -- therefore this wont return None, and then the unwrap below to default to the global will not happen.

This effectively makes the global default ineffective.

Reply:
should be fixed now.